### DISCUSSION TITLE
glTF2+ Support

### DISCUSSION THREAD

**Discussion by HogJonny-AMZN** [o3de/o3de#3750] [Ideas]
This is a page to discussion ideas and various approaches/options for supporting glTF2+ in O3DE.

**Comment by HogJonny-AMZN** [comment]
Option 1: treat glTF as a first class asset type (native support, not import/convert)
(This is just a sketch, may be missing line items.)
- persist glTF data to runtime as much as possible
- load glTF scene data
- load glTF geo and copy to gpu
- implement material/shaders to glTF spec (could re-use shading code from StandardPBR)
- - would need to support gltf schema and material data
- - would need to support packed textures like ORM (R-occlusion, G-Roughness, B-Metalness)
( more TBD )

This option would equate to native runtime support to directly utilize gltf data minimal converting/conditioning to existing O3DE formats.

**Reply by baothientran** [reply]
It maybe also beneficial to support Gltf extensions too, especially for extensions coming from Khronos https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos. Some of those extensions provide support for mesh compression (e.g [KHR_draco_mesh_compression](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_draco_mesh_compression/README.md) and [EXT_meshopt_compression](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Vendor/EXT_meshopt_compression)) and texture compression (e.g [KHR_texture_baisu](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_texture_basisu/README.md) for KTX2) which may help with rendering performance

**Reply by baothientran** [reply]
Also just in case, Khronos provides a lot of gltf samples to test out its spec too https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0

**Reply by HogJonny-AMZN** [reply]
This is what I meant by '2+' we understand people will want to support 'PBR Next' features, etc.

**Reply by HogJonny-AMZN** [reply]
On the Atom side, we have our own asset representations for aspects like geometry, with the data laid out to be performant.
We also already have our own mesh compression.  My understanding is that draco compression is more about compressed size/footprint on disk and optimizations for streaming that data. Which is different from loading that data into the GPU and rendering it.

We also already compress images, typically we take any input texture and based on it's type/usage, we use the most appropriate compression format for the platform it's compressed for.  So we need to consider what this means "KTX v2 images with Basis Universal supercompression. An implementation of this extension can use such images as an alternative to the PNG or JPEG images available in glTF 2.0 for more efficient asset transmission and reducing GPU memory footprint. Furthermore, specifying mip map levels is possible."

The type of questions I have are:
- could we just load a native gltf geo blob and upload to the GPU (I think someone has looked into this already)?
- is that layout (vertex buffers, etc.) somehow better then the one we already have (does this even matter)?
- Is KTX actually universal (like it it a good format for all/any platforms)?
- Is KTX supercompression somehow better then anything we already use?
- Should we add native support for KTX image types, and/or compression (could it be used outside of gltf?)
- etc.

**Reply by amzn-tommy** [reply]
We don't actually have mesh compression right now, but it's something we'll be adding. We also want to move to meshlets with cluster culling. We do currently have some mesh optimization. By having its own native path, we miss out on the opportunity to unify and optimize here. 

We'd be creating a GLTF specific silo if we added full native support. It's not that the silo can't exist, we're just not going to be able to support the same optimizations we do in our standard mesh pipeline unless they are added independently for GLTF. It would be totally possible to have a GLTF gem that used the native GLTF data, a GLTFPBR shader, GLTFMeshComponent, GLTFMeshFeatureProcessor, which submitted its own draw calls. And this might make total sense for a specific application/game that wanted to use a GLTF only pipeline. But to be used interchangeably with any other model, at some point it would have to cross the line into an Atom model, and if we're going to do that, we might as well do it in a builder, unless there is an application that would like to specifically avoid going through the AP before ingesting a GLTF model.

**Reply by Jackie9527** [reply]
> It maybe also beneficial to support Gltf extensions too, especially for extensions coming from Khronos https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos. Some of those extensions provide support for mesh compression (e.g [KHR_draco_mesh_compression](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_draco_mesh_compression/README.md) and [EXT_meshopt_compression](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Vendor/EXT_meshopt_compression)) and texture compression (e.g [KHR_texture_baisu](https://github.com/KhronosGroup/glTF/blob/master/extensions/2.0/Khronos/KHR_texture_basisu/README.md) for KTX2) which may help with rendering performance

assimp supports gltf with draco, building with `-DASSIMP_BUILD_DRACO=ON`.
Lots of gltf files with draco are failed to process by AP.
![image](https://user-images.githubusercontent.com/80555200/222896458-804ee073-0052-44d1-acc8-55c47549c662.png)

**Comment by HogJonny-AMZN** [comment]
Q: Regarding HG's line on the GH issue list for native support, does O3DE have an internal object format/representation that's inclusive of the pieces and parts or does it automatically assign the asset elements to appropriately linked resources across the editor framework?

A: right now no. we do have internal/native representation of all the individual bits and bobs: geo, materials, textures, etc.  The way our AP works is a input (like a .glb), would produce one or more  product outputs.  So we could cook most of the representative data like: a texture, a model (mesh).  But a texture with packed channels would cook, to a .dds, but we don't have a material/shader type (yet) that can do anything with individual channels (we do have configurable texture packing in our backlog.)  We support animations, but the there is no plumbing to pull animations out of gltf and create our runtime version of them.  I 'think' one option for glTF is going to be much like the 'procedural prefab' work up for RFC, an input scene can become an in-engine representation of the whole thing as a prefab.

**Comment by HogJonny-AMZN** [comment]
Shorter Q: is it still one object/asset after import or is it broken up and stored by each type of element?

Good question.  If you flip on support for glTF in the assimp/assetpipeline we probably get 'some' useful data out like geo and probably material assignment slots on the mesh... (revised) looks like some stuff comes through that is supported, like: geo, simple non-texture materials, 'shape keys'/MorphTargets, the skeleton/bones.

- do vertex color work?
- do vertex weights / skinning work?
- can we convert animations?
- what/which material schemas do we support (and how?)
- what rendering features do those scheme need?
- - how much of existing standardPBR shading model / code could be re-used?
- - do we convert from a gltf material to a best-fit atom standardPBR/enahncedPBR? (conversion recipe, map?)
- - or can we make a new .materialType per material spec that uses a similar interface/contract but then includes and utilizes the existing fragment shader, etc.?
- etc.
- etc.

**Reply by amzn-tommy** [reply]
Assimp (as in, the library itself, not our integration) should support all of these. If not, time might be best invested in making sure GLTF works well with Assimp rather than adding native support for GLTF.

An alternative to flipping on GLTF support in Assimp would be to have a GLTF specific builder that imports GLTF->SceneAPI, so that the downstream Atom model builder can continue to consume and optimize things the same way, while making sure GLTF isn't lacking if the Assimp support is found to be lacking and too difficult to fix Assimp itself.

**Comment by HogJonny-AMZN** [comment]
Thought experiment, if I think of it like I might consider USD support in another engine ...

1. Is a native format (I use USD as my internal scene graph)?
2. is it containerized (I persist the USD data, but it's mapped to runtime equivalents transparently)?
3. Is it an import/conversion path (my blender USD/gltf becomes a O3DE prefab, with lots of o3de bits and bobs)?

3) feels like 'procedural prefab' to me, except that if we treat the input asset gltf as 'the source of truth', then if you alter the input gltf you always update the prefab.  And 'procedural prefab' can't be edited the same as a standard prefab (because you break the chain of custody), unless you manually with intend make a detached copy of the 'procedural prefab' which effectively would create a 'standard prefab' copy which could be edited?

4. in/out and roundtrip. If we had number 2 I would want to be able to go back out to USD/gltf

**Comment by mafster** [comment]
Any further progress on importing gltf into O3DE? Still not working it seems. (O3DE 23.05)
